File: Core/conflict_detector.py
"""
Conflict Detection Pipeline

Detects when sources provide contradictory information and
routes to human-in-the-loop resolver for user judgment.

Pipeline:
1. Retrieve multiple sources for query
2. Extract claims from each source
3. Compare claims for contradictions
4. If conflict detected → surface to human
5. Aggregate human judgments to resolve
"""

import logging
from typing import List, Dict, Any, Tuple, Optional
from dataclasses import dataclass
import re

logger = logging.getLogger(__name__)


@dataclass
class ExtractedFact:
    """A single factual claim from a source"""

    fact: str
    subject: str
    predicate: str
    value: str
    source_name: str
    source_type: str
    source_date: Optional[str]
    credibility: float
    relevance: float
    context_snippet: str  # Original sentence for reference


class ConflictDetector:
    """Detect contradictions in multi-source retrieval"""

    def __init__(self):
        self.fact_extraction_patterns = [
            # "X was born in Y"
            r"(\w+[\w\s]*\w+)\s+(?:was\s+)?born\s+(?:in|on)?\s+([^.,;]+)",
            # "X has Y members"
            r"(\w+[\w\s]*\w+)\s+has\s+(\d+[\w\s]*(?:members|employees|residents))",
            # "X is Y"
            r"(\w+[\w\s]*\w+)\s+is\s+([^.,;]+?)(?:\s+and|\s+but|[.,;])",
            # "X founded in Y"
            r"(\w+[\w\s]*\w+)\s+(?:was\s+)?founded\s+(?:in|on)?\s+([^.,;]+)",
            # "X has population of Y"
            r"(\w+[\w\s]*\w+)\s+(?:has\s+)?population\s+(?:of\s+)?(\d+[\w\s,]*)",
        ]

    def detect_conflicts_in_sources(
        self,
        query: str,
        sources: List[Dict[str, Any]],
        min_confidence: float = 0.7,
    ) -> List[Tuple[ExtractedFact, ExtractedFact, str]]:
        """
        Detect factual conflicts across sources.

        Returns:
            List of (fact1, fact2, conflict_summary)
        """
        # Extract facts from all sources
        all_facts = []
        for source in sources:
            facts = self._extract_facts_from_source(source)
            all_facts.extend(facts)

        if len(all_facts) < 2:
            return []

        # Find contradictions
        conflicts = []
        checked = set()

        for i, fact1 in enumerate(all_facts):
            for fact2 in all_facts[i + 1 :]:
                pair_key = (fact1.subject, fact2.subject)

                if pair_key in checked:
                    continue

                # Do they contradict?
                if self._are_contradictory(fact1, fact2):
                    checked.add(pair_key)
                    summary = self._summarize_conflict(fact1, fact2)
                    conflicts.append((fact1, fact2, summary))

        return conflicts

    def _extract_facts_from_source(
        self,
        source: Dict[str, Any],
    ) -> List[ExtractedFact]:
        """Extract factual claims from source text"""
        text = source.get("text", "")
        if not text:
            return []

        facts = []

        for pattern in self.fact_extraction_patterns:
            for match in re.finditer(pattern, text, re.IGNORECASE):
                try:
                    groups = match.groups()
                    if len(groups) >= 2:
                        subject = groups[0].strip()
                        value = groups[1].strip()

                        # Determine predicate from pattern
                        if "born" in pattern:
                            predicate = "born in"
                        elif "population" in pattern:
                            predicate = "has population"
                        elif "founded" in pattern:
                            predicate = "founded in"
                        elif "has" in pattern:
                            predicate = "has"
                        else:
                            predicate = "is"

                        fact = ExtractedFact(
                            fact=f"{subject} {predicate} {value}",
                            subject=subject,
                            predicate=predicate,
                            value=value,
                            source_name=source.get("name", "Unknown"),
                            source_type=source.get("type", "web"),
                            source_date=source.get("date"),
                            credibility=source.get("credibility", 0.5),
                            relevance=source.get("relevance", source.get("score", 0.5)),
                            context_snippet=text[max(0, match.start() - 50) : min(len(text), match.end() + 50)],
                        )

                        facts.append(fact)
                except (IndexError, AttributeError) as e:
                    logger.debug(f"Failed to extract fact from pattern: {e}")

        return facts

    def _are_contradictory(self, fact1: ExtractedFact, fact2: ExtractedFact) -> bool:
        """
        Check if two facts contradict each other.

        Criteria:
        - Same subject
        - Same predicate
        - Different values
        """
        # Must be about the same thing
        subj1 = fact1.subject.lower()
        subj2 = fact2.subject.lower()

        # Strict match OR partial match for longer subjects
        match = subj1 == subj2
        if not match and len(subj1) > 3 and len(subj2) > 3:
            match = (subj1 in subj2) or (subj2 in subj1)

        if not match:
            return False

        # Must be the same predicate
        if fact1.predicate.lower() != fact2.predicate.lower():
            return False

        # Values must be different (not just case/whitespace)
        if fact1.value.lower() == fact2.value.lower():
            return False

        # Check for negation (not true contradiction)
        if self._is_negated(fact1.value) or self._is_negated(fact2.value):
            return False

        return True

    def _is_negated(self, value: str) -> bool:
        """Check if value is negated"""
        negations = ["not", "never", "no ", "none", "isn't", "wasn't", "isn't"]
        return any(neg.lower() in value.lower() for neg in negations)

    def _summarize_conflict(self, fact1: ExtractedFact, fact2: ExtractedFact) -> str:
        """Generate human-readable summary of what conflicts"""
        return f"Sources disagree on '{fact1.predicate}' for '{fact1.subject}': '{fact1.value}' vs '{fact2.value}'"

    def find_supporting_evidence(
        self,
        query: str,
        sources: List[Dict[str, Any]],
        fact_value: str,
        max_results: int = 3,
    ) -> List[Dict[str, Any]]:
        """Find sources that support a particular claim"""
        supporting = []

        for source in sources:
            text = source.get("text", "").lower()
            if fact_value.lower() in text:
                supporting.append(source)
                if len(supporting) >= max_results:
                    break

        return supporting

    def reconcile_with_consensus(
        self,
        conflicts: List[Tuple[ExtractedFact, ExtractedFact, str]],
        all_sources: List[Dict[str, Any]],
    ) -> Dict[str, Any]:
        """
        Try to reconcile conflicts using consensus.

        If N sources agree on a value and only M disagree,
        prefer the consensus but flag as uncertain.
        """
        reconciliation = {}

        for fact1, fact2, summary in conflicts:
            # Count how many sources support each value
            support_1 = self.find_supporting_evidence("", all_sources, fact1.value, max_results=len(all_sources))
            support_2 = self.find_supporting_evidence("", all_sources, fact2.value, max_results=len(all_sources))

            consensus_value = None
            consensus_confidence = 0.5

            if len(support_1) > len(support_2):
                consensus_value = fact1.value
                consensus_confidence = len(support_1) / max(len(support_1) + len(support_2), 1)
            elif len(support_2) > len(support_1):
                consensus_value = fact2.value
                consensus_confidence = len(support_2) / max(len(support_1) + len(support_2), 1)

            reconciliation[summary] = {
                "consensus_value": consensus_value,
                "consensus_confidence": consensus_confidence,
                "count_supporting_1": len(support_1),
                "count_supporting_2": len(support_2),
                "need_human_judgment": (consensus_confidence < 0.66),  # Uncertain if not 2/3 consensus
            }

        return reconciliation

File: rust_output/src/conflict_detector.rs
/// Conflict Detection Pipeline
/// 
/// Detects when sources provide contradictory information and
/// routes to human-in-the-loop resolver for user judgment.
/// 
/// Pipeline:
/// 1. Retrieve multiple sources for query
/// 2. Extract claims from each source
/// 3. Compare claims for contradictions
/// 4. If conflict detected → surface to human
/// 5. Aggregate human judgments to resolve

use anyhow::{Result, Context};
use serde::{Serialize, Deserialize};
use std::collections::HashMap;
use std::collections::HashSet;

pub static LOGGER: std::sync::LazyLock<String /* logging::getLogger */> = std::sync::LazyLock::new(|| Default::default());

/// A single factual claim from a source
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ExtractedFact {
    pub fact: String,
    pub subject: String,
    pub predicate: String,
    pub value: String,
    pub source_name: String,
    pub source_type: String,
    pub source_date: Option<String>,
    pub credibility: f64,
    pub relevance: f64,
    pub context_snippet: String,
}

/// Detect contradictions in multi-source retrieval
#[derive(Debug, Clone)]
pub struct ConflictDetector {
    pub fact_extraction_patterns: Vec<String>,
}

impl ConflictDetector {
    pub fn new() -> Self {
        Self {
            fact_extraction_patterns: vec!["(\\w+[\\w\\s]*\\w+)\\s+(?:was\\s+)?born\\s+(?:in|on)?\\s+([^.,;]+)".to_string(), "(\\w+[\\w\\s]*\\w+)\\s+has\\s+(\\d+[\\w\\s]*(?:members|employees|residents))".to_string(), "(\\w+[\\w\\s]*\\w+)\\s+is\\s+([^.,;]+?)(?:\\s+and|\\s+but|[.,;])".to_string(), "(\\w+[\\w\\s]*\\w+)\\s+(?:was\\s+)?founded\\s+(?:in|on)?\\s+([^.,;]+)".to_string(), "(\\w+[\\w\\s]*\\w+)\\s+(?:has\\s+)?population\\s+(?:of\\s+)?(\\d+[\\w\\s,]*)".to_string()],
        }
    }
    /// Detect factual conflicts across sources.
    /// 
    /// Returns:
    /// List of (fact1, fact2, conflict_summary)
    pub fn detect_conflicts_in_sources(&mut self, query: String, sources: Vec<HashMap<String, Box<dyn std::any::Any>>>, min_confidence: f64) -> Vec<(ExtractedFact, ExtractedFact, String)> {
        // Detect factual conflicts across sources.
        // 
        // Returns:
        // List of (fact1, fact2, conflict_summary)
        let mut all_facts = vec![];
        for source in sources.iter() {
            let mut facts = self._extract_facts_from_source(source);
            all_facts.extend(facts);
        }
        if all_facts.len() < 2 {
            vec![]
        }
        let mut conflicts = vec![];
        let mut checked = HashSet::new();
        for (i, fact1) in all_facts.iter().enumerate().iter() {
            for fact2 in all_facts[(i + 1)..].iter() {
                let mut pair_key = (fact1.subject, fact2.subject);
                if checked.contains(&pair_key) {
                    continue;
                }
                if self._are_contradictory(fact1, fact2) {
                    checked.insert(pair_key);
                    let mut summary = self._summarize_conflict(fact1, fact2);
                    conflicts.push((fact1, fact2, summary));
                }
            }
        }
        conflicts
    }
    /// Extract factual claims from source text
    pub fn _extract_facts_from_source(&mut self, source: HashMap<String, Box<dyn std::any::Any>>) -> Result<Vec<ExtractedFact>> {
        // Extract factual claims from source text
        let mut text = source.get(&"text".to_string()).cloned().unwrap_or("".to_string());
        if !text {
            vec![]
        }
        let mut facts = vec![];
        for pattern in self.fact_extraction_patterns.iter() {
            for r#match in re::finditer(pattern, text, re::IGNORECASE).iter() {
                // try:
                {
                    let mut groups = r#match.groups();
                    if groups.len() >= 2 {
                        let mut subject = groups[0].trim().to_string();
                        let mut value = groups[1].trim().to_string();
                        if pattern.contains(&"born".to_string()) {
                            let mut predicate = "born in".to_string();
                        } else if pattern.contains(&"population".to_string()) {
                            let mut predicate = "has population".to_string();
                        } else if pattern.contains(&"founded".to_string()) {
                            let mut predicate = "founded in".to_string();
                        } else if pattern.contains(&"has".to_string()) {
                            let mut predicate = "has".to_string();
                        } else {
                            let mut predicate = "is".to_string();
                        }
                        let mut fact = ExtractedFact(/* fact= */ format!("{} {} {}", subject, predicate, value), /* subject= */ subject, /* predicate= */ predicate, /* value= */ value, /* source_name= */ source.get(&"name".to_string()).cloned().unwrap_or("Unknown".to_string()), /* source_type= */ source.get(&"type".to_string()).cloned().unwrap_or("web".to_string()), /* source_date= */ source.get(&"date".to_string()).cloned(), /* credibility= */ source.get(&"credibility".to_string()).cloned().unwrap_or(0.5_f64), /* relevance= */ source.get(&"relevance".to_string()).cloned().unwrap_or(source.get(&"score".to_string()).cloned().unwrap_or(0.5_f64)), /* context_snippet= */ text[0.max((r#match.start() - 50))..text.len().min((r#match.end() + 50))]);
                        facts.push(fact);
                    }
                }
                // except (IndexError, AttributeError) as e:
            }
        }
        Ok(facts)
    }
    /// Check if two facts contradict each other.
    /// 
    /// Criteria:
    /// - Same subject
    /// - Same predicate
    /// - Different values
    pub fn _are_contradictory(&mut self, fact1: ExtractedFact, fact2: ExtractedFact) -> bool {
        // Check if two facts contradict each other.
        // 
        // Criteria:
        // - Same subject
        // - Same predicate
        // - Different values
        let mut subj1 = fact1.subject.to_lowercase();
        let mut subj2 = fact2.subject.to_lowercase();
        let mut r#match = subj1 == subj2;
        if (!r#match && subj1.len() > 3 && subj2.len() > 3) {
            let mut r#match = (subj2.contains(&subj1) || subj1.contains(&subj2));
        }
        if !r#match {
            false
        }
        if fact1.predicate.to_lowercase() != fact2.predicate.to_lowercase() {
            false
        }
        if fact1.value.to_lowercase() == fact2.value.to_lowercase() {
            false
        }
        if (self._is_negated(fact1.value) || self._is_negated(fact2.value)) {
            false
        }
        true
    }
    /// Check if value is negated
    pub fn _is_negated(&self, value: String) -> bool {
        // Check if value is negated
        let mut negations = vec!["not".to_string(), "never".to_string(), "no ".to_string(), "none".to_string(), "isn't".to_string(), "wasn't".to_string(), "isn't".to_string()];
        negations.iter().map(|neg| value.to_lowercase().contains(&neg.to_lowercase())).collect::<Vec<_>>().iter().any(|v| *v)
    }
    /// Generate human-readable summary of what conflicts
    pub fn _summarize_conflict(&self, fact1: ExtractedFact, fact2: ExtractedFact) -> String {
        // Generate human-readable summary of what conflicts
        format!("Sources disagree on '{}' for '{}': '{}' vs '{}'", fact1.predicate, fact1.subject, fact1.value, fact2.value)
    }
    /// Find sources that support a particular claim
    pub fn find_supporting_evidence(&self, query: String, sources: Vec<HashMap<String, Box<dyn std::any::Any>>>, fact_value: String, max_results: i64) -> Vec<HashMap<String, Box<dyn std::any::Any>>> {
        // Find sources that support a particular claim
        let mut supporting = vec![];
        for source in sources.iter() {
            let mut text = source.get(&"text".to_string()).cloned().unwrap_or("".to_string()).to_lowercase();
            if text.contains(&fact_value.to_lowercase()) {
                supporting.push(source);
                if supporting.len() >= max_results {
                    break;
                }
            }
        }
        supporting
    }
    /// Try to reconcile conflicts using consensus.
    /// 
    /// If N sources agree on a value and only M disagree,
    /// prefer the consensus but flag as uncertain.
    pub fn reconcile_with_consensus(&mut self, conflicts: Vec<(ExtractedFact, ExtractedFact, String)>, all_sources: Vec<HashMap<String, Box<dyn std::any::Any>>>) -> HashMap<String, Box<dyn std::any::Any>> {
        // Try to reconcile conflicts using consensus.
        // 
        // If N sources agree on a value and only M disagree,
        // prefer the consensus but flag as uncertain.
        let mut reconciliation = HashMap::new();
        for (fact1, fact2, summary) in conflicts.iter() {
            let mut support_1 = self.find_supporting_evidence("".to_string(), all_sources, fact1.value, /* max_results= */ all_sources.len());
            let mut support_2 = self.find_supporting_evidence("".to_string(), all_sources, fact2.value, /* max_results= */ all_sources.len());
            let mut consensus_value = None;
            let mut consensus_confidence = 0.5_f64;
            if support_1.len() > support_2.len() {
                let mut consensus_value = fact1.value;
                let mut consensus_confidence = (support_1.len() / (support_1.len() + support_2.len()).max(1));
            } else if support_2.len() > support_1.len() {
                let mut consensus_value = fact2.value;
                let mut consensus_confidence = (support_2.len() / (support_1.len() + support_2.len()).max(1));
            }
            reconciliation[summary] = HashMap::from([("consensus_value".to_string(), consensus_value), ("consensus_confidence".to_string(), consensus_confidence), ("count_supporting_1".to_string(), support_1.len()), ("count_supporting_2".to_string(), support_2.len()), ("need_human_judgment".to_string(), consensus_confidence < 0.66_f64)]);
        }
        reconciliation
    }
}
